src/test/java/com/commercetools/project/sync/customer/CustomerSyncerTest.java

```diff
@@ -0,0 +1,58 @@
+package com.commercetools.project.sync.customer;
+
+import static io.sphere.sdk.json.SphereJsonUtils.readObjectFromResource;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+import com.commercetools.sync.customers.CustomerSync;
+import com.commercetools.sync.customers.utils.CustomerReferenceResolutionUtils;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.customers.Customer;
+import io.sphere.sdk.customers.CustomerDraft;
+import io.sphere.sdk.customers.queries.CustomerQuery;
+import java.time.Clock;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import org.junit.jupiter.api.Test;
+
+class CustomerSyncerTest {
+  @Test
+  void of_ShouldCreateCustomerSyncerInstance() {
+    // test
+    final CustomerSyncer customerSyncer =
+        CustomerSyncer.of(mock(SphereClient.class), mock(SphereClient.class), mock(Clock.class));
+
+    // assertion
+    assertThat(customerSyncer).isNotNull();
+    assertThat(customerSyncer.getSync()).isInstanceOf(CustomerSync.class);
+  }
+
+  @Test
+  void transform_ShouldReplaceCustomerReferenceIdsWithKeys() {
+    // preparation
+    final CustomerSyncer customerSyncer =
+        CustomerSyncer.of(mock(SphereClient.class), mock(SphereClient.class), mock(Clock.class));
+    final List<Customer> customers =
+        Collections.singletonList(readObjectFromResource("customer-key-1.json", Customer.class));
+
+    // test
+    final CompletionStage<List<CustomerDraft>> draftsFromPageStage =
+        customerSyncer.transform(customers);
+
+    // assertion
+    assertThat(draftsFromPageStage)
+        .isCompletedWithValue(CustomerReferenceResolutionUtils.mapToCustomerDrafts(customers));
+  }
+
+  @Test
+  void getQuery_ShouldBuildCustomerQuery() {
+    // test
+    final CustomerSyncer customerSyncer =
+        CustomerSyncer.of(mock(SphereClient.class), mock(SphereClient.class), mock(Clock.class));
+
+    // assertion
+    final CustomerQuery query = customerSyncer.getQuery();
+    assertThat(query).isEqualTo(CustomerReferenceResolutionUtils.buildCustomerQuery());
+  }
+}
```

Comment:
Can't we include an errorneous workflow to verify that our callbacks are working? I see on productSyncerTest have done mocking a graphQL call. Is it feasible in this syncer unit tests? https://github.com/commercetools/commercetools-project-sync/blob/c849ca06e0794355af7a1f1f3cf2541121ca54be/src/test/java/com/commercetools/project/sync/product/ProductSyncerTest.java#L276

Reply:
Good note. It's definitely possible in our workflow and actually we should do it. If you check my notes in the description, you could see that I wrote this there (https://github.com/commercetools/commercetools-project-sync/pull/184#issue-504625751). 

Unfortunately it's not as straightforward as in for ProductSync. If you look into ProductSyncer, you will see that the GraphQL call happens in the project-sync inside `transform` method, which is actually tested in the test you linked. So you can easily mock that error call as you know that this call will happen for sure when you call `transform()`. 

However, in other syncers there's no ctp call inside `transform()` nor Syncer class. So in order to invoke those callbacks, I would need to mock the error call, but this error call should be inside java-sync as it is the place where callbacks are being invoked. So it means I need to mock much more calls and also I need to mock different error for each resource as every resource call error/warning callback in different cases.

Additionally you might have noticed that the test you linked does NOT test the callbacks, and that's what we're missing here.